source/particle/property_pool.cc

```diff
@@ -0,0 +1,133 @@
+/*
+  Copyright (C) 2016 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+#include <aspect/particle/property_pool.h>
+#include <aspect/particle/particle.h>
+
+#include <aspect/global.h>
+#include <aspect/utilities.h>
+#include <aspect/compat.h>
+
+namespace aspect
+{
+  namespace Particle
+  {
+    PropertyPool::PropertyPool (const unsigned int n_properties_per_slot)
+    :
+          n_properties (n_properties_per_slot),
+          next_free_index (0),
+          next_free_memory_location (0),
+          memory_pool(),
+          translation_table()
+    {}
+
+    PropertyPool::Handle
+    PropertyPool::allocate_properties_array ()
+    {
+      if (n_properties == 0)
+        return Handle();
+
+      if (memory_pool.size() < (next_free_memory_location+1)*n_properties)
+        memory_pool.resize(memory_pool.size() + n_properties);
+
+      if (translation_table.size() < (next_free_index+1))
+        translation_table.resize(translation_table.size() + 1);
```

Comment:
That's really inefficient, because you will resize this all the time. Better multiply the previous size by a fixed factor, say 1.5 or 2.